### PR TITLE
Allow readonly autoprops to be used when generating a constructor

### PR DESCRIPTION
**Pull request by jmarolf**
Fixes 
https://github.com/dotnet/roslyn/issues/13749
https://github.com/dotnet/roslyn/issues/13944
https://github.com/dotnet/roslyn/issues/9022
@dotnet/roslyn-ide

### CODE REVIEW

File: src/Features/Core/Portable/GenerateMember/GenerateConstructor/AbstractGenerateConstructorService.Editor.cs [hunk 1]
```diff
@@ -413,7 +413,6 @@ private IEnumerable<string> GetUnavailableMemberNames()
                         var property = (IPropertySymbol)symbol;
                         return
                             property.Parameters.Length == 0 &&
-                            property.SetMethod != null &&
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/GenerateMember/GenerateConstructor/AbstractGenerateConstructorService.Editor.cs]: we should make sure there's *either* a SetMethod *or* a  backing field.

Can you add a test that if there is an *abstract* getter-only property, that we still don't assign to it?

**Comment by CyrusNajmabadi** [comment]
You could add a general IsWritableInConstructor extension for properties.  That would probably be helpful.

**Comment by CyrusNajmabadi** [comment]
Does this also work for "GenerateConstructorFromMembers"?

**Comment by jmarolf** [comment]
retest vsi please

File: src/EditorFeatures/CSharpTest/Diagnostics/GenerateConstructor/GenerateConstructorTests.cs [hunk 2]
```diff
@@ -3089,5 +3089,16 @@ void M()
 parseOptions: TestOptions.Regular.WithLanguageVersion(CodeAnalysis.CSharp.LanguageVersion.CSharp6),
 withScriptOption: true);
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateConstructor)]
+        [WorkItem(13749, "https://github.com/dotnet/roslyn/issues/13749")]
+        public async Task Support_Readonly_Properties()
+        {
+            await TestAsync(
+@"class C { public int Prop { get ; } } class P { static void M ( ) { var prop = 42 ; var c = new [|C|] ( prop ) ; } } ",
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/GenerateConstructor/GenerateConstructorTests.cs]: no more single line tests :)

File: src/Workspaces/Core/Portable/Shared/Extensions/IPropertySymbolExtensions.cs [hunk 3]
```diff
@@ -69,5 +69,19 @@ public static IPropertySymbol RenameParameters(this IPropertySymbol property, IL
...
                 property.IsIndexer);
         }
+
+        public static bool IsWritableInConstructor(this IPropertySymbol property)
+        {
+            return property.Parameters.Length == 0 &&
+                   (property.SetMethod != null || ContainsBackingField(property));
+        }
+
+        public static IFieldSymbol GetBackingFieldIfAny(this IPropertySymbol property) =>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/IPropertySymbolExtensions.cs]: => on next line.

File: src/Workspaces/Core/Portable/Shared/Extensions/IPropertySymbolExtensions.cs [hunk 4]
```diff
@@ -69,5 +69,19 @@ public static IPropertySymbol RenameParameters(this IPropertySymbol property, IL
...
+            property.ContainingType.GetMembers()
+                                .OfType<IFieldSymbol>()
+                                .FirstOrDefault(f => property.Equals(f.AssociatedSymbol));
+
+        private static bool ContainsBackingField(IPropertySymbol property) =>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/IPropertySymbolExtensions.cs]: => on next line.

File: src/Features/Core/Portable/GenerateMember/GenerateConstructor/AbstractGenerateConstructorService.Editor.cs [hunk 5]
```diff
@@ -412,8 +412,7 @@ private IEnumerable<string> GetUnavailableMemberNames()
                     {
                         var property = (IPropertySymbol)symbol;
                         return
-                            property.Parameters.Length == 0 &&
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/GenerateMember/GenerateConstructor/AbstractGenerateConstructorService.Editor.cs]: keep the .Parameters check here.

File: src/Workspaces/Core/Portable/Shared/Extensions/IPropertySymbolExtensions.cs [hunk 6]
```diff
@@ -69,5 +69,19 @@ public static IPropertySymbol RenameParameters(this IPropertySymbol property, IL
                 property.SetMethod,
                 property.IsIndexer);
         }
+
+        public static bool IsWritableInConstructor(this IPropertySymbol property)
+        {
+            return property.Parameters.Length == 0 &&
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/IPropertySymbolExtensions.cs]: a property is still assignable even if Parameters.Length is > 0.

File: src/Workspaces/Core/Portable/Shared/Extensions/IPropertySymbolExtensions.cs (hunk 3 above)
Comment (jmarolf) [inline src/Workspaces/Core/Portable/Shared/Extensions/IPropertySymbolExtensions.cs]: Like this?

```
public static IFieldSymbol GetBackingFieldIfAny(this IPropertySymbol property)
    => property.ContainingType.GetMembers()
        .OfType<IFieldSymbol>()
        .FirstOrDefault(f => property.Equals(f.AssociatedSymbol));
```

---
In reply to: [86491171](https://github.com/dotnet/roslyn/pull/14827#discussion_r86491171) [](ancestors = 86491171)

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/IPropertySymbolExtensions.cs]: yup!

**Comment by CyrusNajmabadi** [comment]
👍

**Pull request merged**